### PR TITLE
Bring intial changes into master

### PR DESCRIPTION
**Pull request by slashrsm**

### CODE REVIEW

File: src/hdprivatekey.js [hunk 1]
```diff
@@ -547,14 +547,6 @@ HDPrivateKey.prototype.toString = function() {
   return this.xprivkey;
 };
 
-/**
- * Returns the console representation of this extended private key.
- * @return string
- */
-HDPrivateKey.prototype.inspect = function() {
```
Comment (slashrsm) [inline src/hdprivatekey.js]: Just curious what was the reason for this removal? Sure, console.log() in production code isn't the best thing ever, but this could stay I suppose?

File: src/transaction/input/publickey.js [hunk 2]
```diff
@@ -29,9 +29,9 @@ inherits(PublicKeyInput, Input);
  * @return {Array} of objects that can be
  */
 PublicKeyInput.prototype.getSignatures = function(transaction, privateKey, index, sigtype) {
-  $.checkState(this.output instanceof Output);
+  $.checkState(this.output instanceof Output, 'Malformed output found when signing transaction');
   sigtype = sigtype || (Signature.SIGHASH_ALL |  Signature.SIGHASH_FORKID);
-  var publicKey = privateKey.toPublicKey();
+  var publicKey = privateKey.toPublicKey()
```
Comment (slashrsm) [inline src/transaction/input/publickey.js]: Was this change intended?

File: src/transaction/input/scripthash.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ * @constructor
+ */
+function ScriptHashInput(input, pubkeys, redeemScript) {
+  Input.apply(this, arguments);
+  var self = this;
+  this.publicKeys = pubkeys || input.publicKeys;
+  this.threshold = 1;
+  this.redeemScript = redeemScript;
+  $.checkState(Script.buildScriptHashOut(this.redeemScript).equals(this.output.script),
+    'Provided redeemScript doesn\'t hash to the provided output');
```
Comment (slashrsm) [inline src/transaction/input/scripthash.js]: We [should use backslashes](https://github.com/airbnb/javascript#strings--escaping) to delimit the string and avoid escaping.

File: src/transaction/input/scripthash.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+ScriptHashInput.prototype._serializeSignatures = function() {
+  return this.signatures.map(signature => signature ? signature.toObject() : undefined);
+};
+
+ScriptHashInput.prototype.getSignatures = function(transaction, privateKey, index, sigtype) {
+  $.checkState(this.output instanceof Output, 'Malformed output found when signing transaction');
+  sigtype = sigtype || (Signature.SIGHASH_ALL |  Signature.SIGHASH_FORKID);
+
+  var self = this;
+  var results = [];
```
Comment (slashrsm) [inline src/transaction/input/scripthash.js]: Above we converted a similar hunk into something a bit more functionally oriented (filter(), ...). Should we use the same pattern here?

File: src/transaction/input/scripthash.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  return this.signatures.reduce((sum, signature) => sum + !!signature, 0);
+};
+
+ScriptHashInput.prototype.publicKeysWithoutSignature = function() {
+  var self = this;
+  return this.publicKeys.filter(publicKey => !(self.signatures[self.publicKeyIndex[publicKey.toString()]]));
+};
+
+ScriptHashInput.prototype.isValidSignature = function(transaction, signature) {
+  // FIXME: Refactor signature so this is not necessary
```
Comment (slashrsm) [inline src/transaction/input/scripthash.js]: Follow-up?

File: src/hdprivatekey.js (hunk 1 above)
Comment (ClemensLey) [reply src/hdprivatekey.js]: I was getting a warning that inspect is deprecated
https://nodejs.org/api/deprecations.html#deprecations_dep0079_custom_inspection_function_on_objects_via_inspect

There would have been a more consistent way to fix but I was preparing for a demo and wanted a quick fix. We should think about what to do with the inspect functions that are in almost all classes

File: src/transaction/input/publickey.js (hunk 2 above)
Comment (ClemensLey) [reply src/transaction/input/publickey.js]: probably not

File: src/transaction/input/scripthash.js (hunk 4 above)
Comment (ClemensLey) [reply src/transaction/input/scripthash.js]: I think so

File: src/hdprivatekey.js (hunk 1 above)
Comment (slashrsm) [reply src/hdprivatekey.js]: Created followup: #11

File: src/transaction/input/scripthash.js (hunk 5 above)
Comment (slashrsm) [reply src/transaction/input/scripthash.js]: created: #13 12

**Comment by slashrsm** [comment]
- Fixed cosmetic comments
- Created followups for test coverage, figuring out inspect and refactor of isValidSignature
- Squashed and rebased commit history
